### PR TITLE
rename context to context_autoregress for inference

### PR DESCRIPTION
**Pull request by gobbleturk**
# Description

This is a copy of Qinwen's https://github.com/AI-Hypercomputer/maxtext/pull/1453 - she is OOO for a while and we want this merged soon so I've copied it here

rename context_parallelism to context_autoregressive_parallelism to separate CP naming for inference

The rest of the description includes relevant details and context, examples:

training and inference have different sharding stategies to best performance, like to fdsp qkv weights or not.
introduce context_autoregressive_parallelism as cp axes for inference only.
If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/405621754

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/linears.py [hunk 1]
```diff
@@ -852,7 +852,7 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
         # todo: try replace softmax_probs with padded weights and verify with decode acc tests
         softmax_probs = jax.nn.softmax(gate_logits.astype(jnp.float32), axis=-1).astype(self.dtype)
         dispatch_mask, combine_mask = self.generate_masks_subgroup(top_k_indices, softmax_probs)
-        if self.config.ici_context_parallelism > 0 and cp == 1:
+        if self.config.ici_context_autoregressive_parallelism > 1 and cp == 1:
```
Comment (richjames0) [inline MaxText/layers/linears.py:855]: maybe comment this change

Comment (richjames0) [reply MaxText/layers/linears.py:855]: also a unit test

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -1035,13 +1033,12 @@ def qkv_projection(self, inputs: Array, proj_name: str):
     query, key, value = qkv_proj[:, :, 0, ...], qkv_proj[:, :, 1, ...], qkv_proj[:, :, 2, ...]
     return query, key, value
 
-  def out_projection(self, output_dim: int, out: Array) -> Array:
-    out_kernel_axis = (None, None, None) if self.config.ici_context_parallelism > 1 else ("heads", "kv", "embed")
+  def out_projection(self, output_dim: int, out: Array, kernel_axes: AxisNames = ("heads", "kv", "embed")) -> Array:
```
Comment (richjames0) [inline MaxText/layers/attentions.py:1036]: can we get a unit test on this?

File: MaxText/layers/attentions.py [hunk 3]
```diff
@@ -985,7 +984,9 @@ def query_init(*args):
     )(inputs_q)
     return query_proj
 
-  def kv_projection(self, inputs_kv: Array, proj_name: str) -> Array:
+  def kv_projection(
```
Comment (richjames0) [inline MaxText/layers/attentions.py:987]: unit test?

File: MaxText/layers/attentions.py [hunk 4]
```diff
@@ -959,7 +959,7 @@ def setup(self):
           attn_logits_soft_cap=self.attn_logits_soft_cap,
       )
 
-  def query_projection(self, inputs_q: Array) -> Array:
+  def query_projection(self, inputs_q: Array, kernel_axes: AxisNames = ("embed", "q_heads", "kv")) -> Array:
```
Comment (richjames0) [inline MaxText/layers/attentions.py:962]: unit test?

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -322,7 +322,7 @@ logical_axis_rules: [
                       ['paged_kv_head_dim_size', []],
                     ]
 # Axes used for DCN must be earlier in this list than ICI, see (b/339009148) for details
-data_sharding: [['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']]
+data_sharding: [['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive', 'context_autoregressive']]
```
Comment (richjames0) [inline MaxText/configs/base.yml:325]: is this change intentional

**Review by richjames0 — COMMENTED** [review]
The name of this PR seems inaccurate - there are changes going on beyond renaming. Would in general love for us to start unit testing things like this :)

**Comment by gobbleturk** [comment]
Good points rich, I will close this one in favor of a pure name refactor https://github.com/AI-Hypercomputer/maxtext/pull/1501

File: MaxText/maxtext_utils.py [hunk 6]
```diff
@@ -313,7 +313,6 @@ def assert_params_sufficiently_sharded(params, mesh, tolerance):
       "fsdp",
       "fsdp_transpose",
       "sequence",
-      "context",
```
Comment (bvandermoon) [inline MaxText/maxtext_utils.py:316]: Why removing from here instead of renaming?

**Review by bvandermoon — COMMENTED** [review]
Generally LGTM pending the outcome of @richjames0's comments
